### PR TITLE
FolderManager: Rename before create new file

### LINKED ISSUE #213: Creating new file immediately opens it
When creating a new file in the sidebar, Scratch shouldn't immediately open it in the editor with the "unnamed" name, but allow for renaming it, and after the renaming is done, the file should be loaded in the editor.

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/49059093-creating-new-file-immediately-opens-it?utm_campaign=plugin&utm_content=tracker%2F61917289&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F61917289&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by jeremypw**
Fixes #213 

An alternative to #650, which should be less prone to race conditions.  The filename to be created is entered *before* any file is created using a temporary renaming item.  To match Files behaviour, if <Esc> is pressed (or a blank name entered) then a file with the default name is created rather than cancelling the file creation.

It proved necessary to poll the SourceList to determine when editing finishes because cancelling the editing does not generate any signal and we need to remove the temporary item in either case.

If this were fixed in Granite then a simplification here would be possible.

### CODE REVIEW

File: src/FolderManager/FolderItem.vala [hunk 1]
```diff
@@ -328,67 +320,90 @@ namespace Scratch.FolderManager {
...
+        private void on_add_new (bool is_folder) {
             if (!file.is_executable) {
                 // This is necessary to avoid infinite loop below
                 warning ("Unable to open parent folder");
                 return;
             }
 
-            var new_folder = file.file.get_child (_("untitled folder"));
-
+            var name = is_folder ? _("untitled folder") : _("new file");
```
Comment (tintou) [inline src/FolderManager/FolderItem.vala]:
```suggestion
            unowned string name = is_folder ? _("untitled folder") : _("new file");
```

File: src/FolderManager/FolderItem.vala [hunk 2]
```diff
@@ -328,67 +320,90 @@ namespace Scratch.FolderManager {
...
+
+                /* Need to poll view editing as no signal is generated when canceled (Granite bug) */
+                Timeout.add (200, () => {
+                    if (view.editing) {
+                        return Source.CONTINUE;
+                    } else {
+                        var new_name = rename_item.name;
+                        remove (rename_item);
+                        var path = file.file.get_path () + "/" + new_name;
+                        GLib.File gfile = GLib.File.new_for_path (path);
```
Comment (tintou) [inline src/FolderManager/FolderItem.vala]:
```suggestion
                        var gfile = file.file.get_child_for_display_name (rename_item.name);
                        remove (rename_item);
```

File: src/FolderManager/FolderItem.vala [hunk 3]
```diff
@@ -328,67 +320,90 @@ namespace Scratch.FolderManager {
...
+        }
 
-                if (!children_loaded) {
-                    clear ();
-                    add_children ();
-                    children_loaded = true;
-                }
-            } catch (Error e) {
-                warning (e.message);
+        protected new void rename (string new_name) {
```
Comment (tintou) [inline src/FolderManager/FolderItem.vala]:
```suggestion
        protected new void rename (string new_name) {
```
this should really be avoided, I really don't like the use of the `new` keyword.

File: src/FolderManager/FolderItem.vala (hunk 1 above)
Comment (jeremypw) [reply src/FolderManager/FolderItem.vala]: I did not think that there would be any string copying in this context (assigning from an expression not a variable or property)?  But I'll follow your advice.

File: src/FolderManager/FolderItem.vala (hunk 3 above)
Comment (jeremypw) [reply src/FolderManager/FolderItem.vala]: No difference in you suggested line?  I'll see if I can find an alternative.

**Comment by jeremypw** [comment]
@tintou Thanks for the review :-)   Do you think this a workable approach compared to the alternative PR (which I found difficult to debug)?

File: src/FolderManager/FolderItem.vala (hunk 3 above)
Comment (jeremypw) [reply src/FolderManager/FolderItem.vala]: On further inspection, new is not needed - I was confused between FolderManager.Item (which was the initial base class I used) and Granite.Widgets.SourceList.Item.

File: src/FolderManager/FolderItem.vala (hunk 2 above)
Comment (jeremypw) [reply src/FolderManager/FolderItem.vala]: OK

Comment (tintou) [reply src/FolderManager/FolderItem.vala]: have you seen this, it looks safer and more efficient to what you're doing :)

File: src/FolderManager/FolderItem.vala (hunk 1 above)
Comment (tintou) [reply src/FolderManager/FolderItem.vala]: `var name = "foo"` results to `name = g_strdup ("foo");` in C when `unowned string name = "foo"` results in `name = "foo"` so it's a tiny gain here ;)

**Comment by tintou** [comment]
Yeah it looks really better, hopefully once Gtk4 will be out Granite Source View won't be a thing anymore :slightly_smiling_face:

File: src/FolderManager/FolderItem.vala (hunk 2 above)
Comment (jeremypw) [reply src/FolderManager/FolderItem.vala]: OOps - I just saw the change to var.  I'll use your improved version.

**Comment by jeremypw** [comment]
Missed change made - sorry about that.

**Pull request merged**